File: crates/unottr-core/src/media/ffmpeg.rs
use std::env;
use std::io::{BufRead, BufReader, Read};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use serde::Deserialize;
use tracing::{debug, warn};

use super::{AudioStream, MediaBackend, Probe, TARGET_SAMPLE_RATE};
use crate::cancel::CancelToken;
use crate::error::{Error, IoResultExt, Result};

/// `current_exe`'s directory is checked for `name`; falls back to the bare name so
/// `Command` resolves it against `PATH` instead.
fn bundled_or(name: &str) -> PathBuf {
    if let Ok(exe) = env::current_exe()
        && let Some(dir) = exe.parent()
    {
        let candidate = dir.join(name);
        if candidate.is_file() {
            return candidate;
        }
    }
    name.into()
}

/// Drives the system ffmpeg/ffprobe binaries.
#[derive(Debug, Clone)]
pub struct FfmpegCli {
    ffmpeg: PathBuf,
    ffprobe: PathBuf,
}

impl Default for FfmpegCli {
    fn default() -> Self {
        Self::discover()
    }
}

impl FfmpegCli {
    /// Priority: explicit env override, then a binary bundled next to the running
    /// executable (how `scripts/package-tarball.sh`'s fallback ships ffmpeg; the AppImage
    /// path instead points the env vars at its resource dir at startup, see src-tauri's
    /// `run()`), then bare name for the OS to resolve against `PATH`. No Tauri knowledge
    /// needed here — packaging just has to get `ffmpeg`/`ffprobe` in front of one of these.
    pub fn discover() -> Self {
        Self {
            ffmpeg: env::var_os("UNOTTR_FFMPEG")
                .map_or_else(|| bundled_or("ffmpeg"), PathBuf::from),
            ffprobe: env::var_os("UNOTTR_FFPROBE")
                .map_or_else(|| bundled_or("ffprobe"), PathBuf::from),
        }
    }

    pub fn new(ffmpeg: impl Into<PathBuf>, ffprobe: impl Into<PathBuf>) -> Self {
        Self {
            ffmpeg: ffmpeg.into(),
            ffprobe: ffprobe.into(),
        }
    }

    /// Both settings must be set to override `discover()` — leaves `UNOTTR_FFMPEG`/`PATH`
    /// lookup intact for the common case of overriding neither. Shared by the tauri app's
    /// startup backend construction and its ffmpeg-preflight check so the two never drift.
    pub fn from_settings(ffmpeg_path: Option<&str>, ffprobe_path: Option<&str>) -> Self {
        match (ffmpeg_path, ffprobe_path) {
            (Some(f), Some(p)) if !f.is_empty() && !p.is_empty() => Self::new(f, p),
            _ => Self::discover(),
        }
    }

    /// Both binaries resolve and run. Surfaced in settings so the user sees the problem
    /// before a recording fails rather than after.
    pub fn check(&self) -> Result<()> {
        for bin in [&self.ffmpeg, &self.ffprobe] {
            let status = Command::new(bin)
                .arg("-version")
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .status()
                .map_err(|e| map_spawn(bin, e))?;
            if !status.success() {
                return Err(Error::Ffmpeg(format!("{} -version failed", bin.display())));
            }
        }
        Ok(())
    }
}

impl MediaBackend for FfmpegCli {
    fn probe(&self, path: &Path) -> Result<Probe> {
        // ffprobe answers "invalid data" for a missing file too; stat first so the user gets
        // the real reason
        std::fs::metadata(path).at(path)?;

        let out = Command::new(&self.ffprobe)
            .args([
                "-v",
                "error",
                "-print_format",
                "json",
                "-show_format",
                "-show_streams",
            ])
            .arg(path)
            .output()
            .map_err(|e| map_spawn(&self.ffprobe, e))?;

        if !out.status.success() {
            let stderr = String::from_utf8_lossy(&out.stderr);
            // mp4 writes its index last, so an interrupted recording has no moov atom. this
            // is the ordinary "obs died mid-meeting" case, not a corrupt file.
            return Err(if stderr.contains("moov atom not found") {
                Error::Truncated {
                    path: path.to_path_buf(),
                }
            } else {
                Error::Probe {
                    path: path.to_path_buf(),
                    reason: stderr
                        .lines()
                        .find(|l| !l.trim().is_empty())
                        .unwrap_or("ffprobe failed")
                        .trim()
                        .to_string(),
                }
            });
        }

        let raw: RawProbe = serde_json::from_slice(&out.stdout).map_err(|e| Error::Probe {
            path: path.to_path_buf(),
            reason: format!("unreadable ffprobe output: {e}"),
        })?;

        let mut audio = Vec::new();
        let mut has_video = false;
        for stream in &raw.streams {
            match stream.codec_type.as_str() {
                // attached cover art is a video stream by ffprobe's reckoning; ignore it
                "video" if stream.disposition.attached_pic == 0 => has_video = true,
                "audio" => {
                    let audio_index = audio.len() as u32;
                    audio.push(AudioStream {
                        index: stream.index,
                        audio_index,
                        codec: stream.codec_name.clone().unwrap_or_else(|| "?".into()),
                        channels: stream.channels.unwrap_or(0),
                        sample_rate: stream
                            .sample_rate
                            .as_deref()
                            .and_then(|r| r.parse().ok())
                            .unwrap_or(0),
                        title: stream.tags.get("title").cloned(),
                        language: stream.tags.get("language").cloned(),
                    });
                }
                _ => {}
            }
        }

        if audio.is_empty() {
            return Err(Error::NoAudio {
                path: path.to_path_buf(),
            });
        }

        // a still-recording or interrupted file reports no duration; treating that as
        // truncated is what lets the watcher wait instead of transcribing a partial file
        let duration_ms = raw
            .format
            .as_ref()
            .and_then(|f| f.duration.as_deref())
            .and_then(|d| d.parse::<f64>().ok())
            .filter(|d| d.is_finite() && *d > 0.0)
            .map(|d| (d * 1000.0).round() as u64);
        if duration_ms.is_none() {
            return Err(Error::Truncated {
                path: path.to_path_buf(),
            });
        }

        Ok(Probe {
            container: raw
                .format
                .as_ref()
                .map_or_else(|| "?".into(), |f| f.format_name.clone()),
            duration_ms,
            has_video,
            audio,
        })
    }

    fn extract_pcm(
        &self,
        path: &Path,
        audio_index: u32,
        out: &Path,
        progress: &dyn Fn(f32),
        cancel: &CancelToken,
    ) -> Result<()> {
        cancel.check()?;
        let total_ms = self.probe(path)?.duration_ms.unwrap_or(0);

        if let Some(parent) = out.parent() {
            std::fs::create_dir_all(parent).at(parent)?;
        }

        let mut child = Command::new(&self.ffmpeg)
            .args(["-nostdin", "-v", "error", "-progress", "pipe:1", "-y", "-i"])
            .arg(path)
            .args([
                "-map",
                &format!("0:a:{audio_index}"),
                "-vn",
                "-sn",
                "-dn",
                "-ac",
                "1",
                "-ar",
                &TARGET_SAMPLE_RATE.to_string(),
                "-acodec",
                "pcm_s16le",
                "-f",
                "s16le",
            ])
            .arg(out)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| map_spawn(&self.ffmpeg, e))?;

        // stderr must be drained concurrently or ffmpeg blocks on a full pipe and we deadlock
        let mut stderr = child.stderr.take().expect("stderr piped");
        let drain = std::thread::spawn(move || {
            let mut buf = String::new();
            let _ = stderr.read_to_string(&mut buf);
            buf
        });

        let stdout = BufReader::new(child.stdout.take().expect("stdout piped"));
        let mut cancelled = false;
        for line in stdout.lines() {
            let Ok(line) = line else { break };
            if let Some(done_ms) = progress_ms(&line)
                && total_ms > 0
            {
                progress((done_ms as f32 / total_ms as f32).clamp(0.0, 1.0));
            }
            if cancel.is_cancelled() {
                cancelled = true;
                let _ = child.kill();
                break;
            }
        }

        let status = child.wait().map_err(|e| map_spawn(&self.ffmpeg, e))?;
        let stderr = drain.join().unwrap_or_default();

        if cancelled {
            discard(out);
            return Err(Error::Cancelled);
        }
        if !status.success() {
            discard(out);
            let tail = stderr.trim().lines().next_back().unwrap_or("no output");
            return Err(Error::Ffmpeg(format!("{}: {tail}", path.display())));
        }

        progress(1.0);
        debug!(file = %path.display(), track = audio_index, "extracted pcm");
        Ok(())
    }
}

/// ffmpeg's `-progress` stream is `key=value` lines; `out_time_us` is authoritative,
/// `out_time_ms` is a misnomer that also holds microseconds on older builds.
fn progress_ms(line: &str) -> Option<u64> {
    let (key, value) = line.split_once('=')?;
    match key {
        "out_time_us" | "out_time_ms" => value.trim().parse::<u64>().ok().map(|us| us / 1000),
        _ => None,
    }
}

/// A partial pcm is worse than none: phase 04 treats an existing file as a finished one.
fn discard(out: &Path) {
    if let Err(e) = std::fs::remove_file(out)
        && e.kind() != std::io::ErrorKind::NotFound
    {
        warn!(file = %out.display(), error = %e, "could not remove partial pcm");
    }
}

fn map_spawn(bin: &Path, e: std::io::Error) -> Error {
    if e.kind() == std::io::ErrorKind::NotFound {
        Error::FfmpegMissing
    } else {
        Error::Io {
            path: bin.to_path_buf(),
            source: e,
        }
    }
}

#[derive(Deserialize)]
struct RawProbe {
    #[serde(default)]
    streams: Vec<RawStream>,
    format: Option<RawFormat>,
}

#[derive(Deserialize)]
struct RawStream {
    index: u32,
    codec_type: String,
    codec_name: Option<String>,
    channels: Option<u16>,
    sample_rate: Option<String>,
    #[serde(default)]
    tags: std::collections::HashMap<String, String>,
    #[serde(default)]
    disposition: RawDisposition,
}

#[derive(Deserialize, Default)]
struct RawDisposition {
    #[serde(default)]
    attached_pic: u8,
}

#[derive(Deserialize)]
struct RawFormat {
    format_name: String,
    duration: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_microseconds_from_the_progress_stream() {
        assert_eq!(progress_ms("out_time_us=1500000"), Some(1500));
        assert_eq!(progress_ms("out_time_ms=1500000"), Some(1500));
        assert_eq!(progress_ms("frame=12"), None);
        assert_eq!(progress_ms("out_time_us=N/A"), None);
        assert_eq!(progress_ms("progress=continue"), None);
    }

    #[test]
    fn a_missing_file_is_reported_as_such() {
        let err = FfmpegCli::discover()
            .probe(Path::new("/nonexistent-unottr.mkv"))
            .unwrap_err();
        assert!(matches!(err, Error::Io { .. }), "got {err}");
    }

    #[test]
    fn a_missing_binary_is_not_an_io_error() {
        let cli = FfmpegCli::new("unottr-no-such-ffmpeg", "unottr-no-such-ffprobe");
        assert!(matches!(cli.check(), Err(Error::FfmpegMissing)));
        // a file that exists, so we get past the stat and actually try to spawn
        assert!(matches!(
            cli.probe(Path::new("/etc/hostname")),
            Err(Error::FfmpegMissing)
        ));
    }

    fn have_ffmpeg() -> bool {
        FfmpegCli::discover().check().is_ok()
    }

    /// 07-hardening-and-packaging.md's robustness pass: hostile inputs must produce a typed
    /// error, never a panic. A 0-byte file has no moov atom either, so most ffprobe builds
    /// report it the same way as a truncated recording; accept either typed outcome since
    /// the exact wording isn't the point.
    #[test]
    fn a_zero_byte_file_is_a_typed_error_not_a_panic() {
        if !have_ffmpeg() {
            return;
        }
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("empty.mp4");
        std::fs::File::create(&path).unwrap();
        let err = FfmpegCli::discover().probe(&path).unwrap_err();
        assert!(matches!(err, Error::Probe { .. } | Error::Truncated { .. }), "got {err}");
    }

    /// Plain text with a media extension — ffprobe can't sniff a container out of it at all.
    #[test]
    fn a_text_file_disguised_as_mkv_is_a_typed_error_not_a_panic() {
        if !have_ffmpeg() {
            return;
        }
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("notes.mkv");
        std::fs::write(&path, b"just some plain text, not a container at all\n").unwrap();
        let err = FfmpegCli::discover().probe(&path).unwrap_err();
        assert!(matches!(err, Error::Probe { .. } | Error::Truncated { .. }), "got {err}");
    }

    /// Simulates the file vanishing between discovery and extraction (watch folder on a
    /// removable drive, user deletes the source, etc). `extract_pcm` re-probes internally, so
    /// this exercises the same vanished-file path as a mid-pipeline disappearance would.
    #[test]
    fn a_file_deleted_before_extraction_is_a_typed_error_not_a_panic() {
        if !have_ffmpeg() {
            return;
        }
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("vanishing.mp4");
        std::fs::write(&path, b"not real media, just bytes").unwrap();
        let out = dir.path().join("out.pcm");
        let cancel = CancelToken::new();
        std::fs::remove_file(&path).unwrap();

        let err = FfmpegCli::discover()
            .extract_pcm(&path, 0, &out, &|_| {}, &cancel)
            .unwrap_err();
        assert!(matches!(err, Error::Io { .. } | Error::Probe { .. } | Error::Ffmpeg(_)), "got {err}");
        assert!(!out.exists());
    }
}

File: src/main/media/ffmpeg.ts
// Port of crates/unottr-core/src/media/ffmpeg.rs. Strictly read-only on source files:
// ffmpeg only ever writes the pcm we name.

import { spawn, spawnSync } from "node:child_process";
import { existsSync, renameSync, statSync, unlinkSync } from "node:fs";
import { mkdir } from "node:fs/promises";
import { dirname, join } from "node:path";
import { err, isCancelled } from "../errors";
import type { AudioStream, Probe } from "./types";
import { TARGET_SAMPLE_RATE } from "./types";

export interface FfmpegCli {
  ffmpeg: string;
  ffprobe: string;
}

export function executableName(
  name: string,
  platform: NodeJS.Platform = process.platform,
): string {
  return platform === "win32" && !name.toLowerCase().endsWith(".exe") ? `${name}.exe` : name;
}

/** platform resource -> legacy resource -> PATH */
export function bundledCandidates(
  name: string,
  host: {
    platform?: NodeJS.Platform;
    arch?: string;
    resourcesPath?: string;
    execPath?: string;
    cwd?: string;
  } = {},
): string[] {
  const platform = host.platform ?? process.platform;
  const platformDir = `${platform}-${host.arch ?? process.arch}`;
  const file = executableName(name, platform);
  const resourcesPath = host.resourcesPath ?? process.resourcesPath;
  const bases = [
    resourcesPath ? join(resourcesPath, "bin") : null,
    join(dirname(host.execPath ?? process.execPath), "bin"),
    join(host.cwd ?? process.cwd(), "resources", "bin"),
  ].filter((dir): dir is string => dir !== null);
  return bases.flatMap((dir) => [join(dir, platformDir, file), join(dir, file)]);
}

function bundledOr(name: string): string {
  for (const candidate of bundledCandidates(name)) {
    if (existsSync(candidate)) return candidate;
  }
  return name;
}

export function discover(): FfmpegCli {
  return {
    ffmpeg: process.env.UNOTTR_FFMPEG || bundledOr("ffmpeg"),
    ffprobe: process.env.UNOTTR_FFPROBE || bundledOr("ffprobe"),
  };
}

/** Both settings must be set to override discovery — overriding neither is the common case. */
export function fromSettings(ffmpegPath?: string | null, ffprobePath?: string | null): FfmpegCli {
  return ffmpegPath && ffprobePath ? { ffmpeg: ffmpegPath, ffprobe: ffprobePath } : discover();
}

/** Both binaries resolve and run. Shown in settings so the user sees this before a job fails. */
export function check(cli: FfmpegCli = discover()): boolean {
  return [cli.ffmpeg, cli.ffprobe].every((bin) => {
    const r = spawnSync(bin, ["-version"], { stdio: "ignore" });
    return !r.error && r.status === 0;
  });
}

export async function probe(cli: FfmpegCli, path: string): Promise<Probe> {
  // ffprobe answers "invalid data" for a missing file too; stat first so the user gets the
  // real reason
  try {
    statSync(path);
  } catch (e) {
    throw err.io(path, e);
  }

  const out = await run(cli.ffprobe, [
    "-v", "error", "-print_format", "json", "-show_format", "-show_streams", path,
  ]);

  if (out.code !== 0) {
    // mp4 writes its index last, so an interrupted recording has no moov atom. this is the
    // ordinary "obs died mid-meeting" case, not a corrupt file.
    if (out.stderr.includes("moov atom not found")) throw err.truncated(path);
    const line = out.stderr.split("\n").find((l) => l.trim() !== "");
    throw err.probe(path, (line ?? "ffprobe failed").trim());
  }

  let raw: RawProbe;
  try {
    raw = JSON.parse(out.stdout) as RawProbe;
  } catch (e) {
    throw err.probe(path, `unreadable ffprobe output: ${e instanceof Error ? e.message : e}`);
  }

  const audio: AudioStream[] = [];
  let hasVideo = false;
  for (const s of raw.streams ?? []) {
    // attached cover art is a video stream by ffprobe's reckoning; ignore it
    if (s.codec_type === "video" && (s.disposition?.attached_pic ?? 0) === 0) hasVideo = true;
    if (s.codec_type !== "audio") continue;
    audio.push({
      index: s.index,
      audio_index: audio.length,
      codec: s.codec_name ?? "?",
      channels: s.channels ?? 0,
      sample_rate: int(s.sample_rate) ?? 0,
      title: s.tags?.title ?? null,
      language: s.tags?.language ?? null,
    });
  }
  if (audio.length === 0) throw err.noAudio(path);

  // a still-recording or interrupted file reports no duration; treating that as truncated is
  // what lets the watcher wait instead of transcribing a partial file
  const seconds = Number(raw.format?.duration);
  if (!Number.isFinite(seconds) || seconds <= 0) throw err.truncated(path);

  return {
    container: raw.format?.format_name ?? "?",
    duration_ms: Math.round(seconds * 1000),
    has_video: hasVideo,
    audio,
  };
}

export interface ExtractOptions {
  /** Position among audio streams, not the global stream index — this is `-map 0:a:N`. */
  audioIndex: number;
  out: string;
  onProgress?: (pct: number) => void;
  signal?: AbortSignal;
}

/** Decodes one audio track to 16 kHz mono s16le. */
export async function extractPcm(cli: FfmpegCli, path: string, o: ExtractOptions): Promise<void> {
  throwIfAborted(o.signal);
  const totalMs = (await probe(cli, path)).duration_ms ?? 0;
  await mkdir(dirname(o.out), { recursive: true });

  const args = [
    "-nostdin", "-v", "error", "-progress", "pipe:1", "-y", "-i", path,
    "-map", `0:a:${o.audioIndex}`, "-vn", "-sn", "-dn",
    "-ac", "1", "-ar", String(TARGET_SAMPLE_RATE), "-acodec", "pcm_s16le", "-f", "s16le",
    o.out,
  ];

  const result = await run(cli.ffmpeg, args, {
    signal: o.signal,
    onStdoutLine: (line) => {
      const doneMs = progressMs(line);
      if (doneMs !== null && totalMs > 0) {
        o.onProgress?.(Math.min(1, Math.max(0, doneMs / totalMs)));
      }
    },
  });

  if (result.aborted) {
    discard(o.out);
    throw err.cancelled();
  }
  if (result.code !== 0) {
    discard(o.out);
    const lines = result.stderr.trim().split("\n").filter((l) => l !== "");
    throw err.ffmpeg(`${path}: ${lines.at(-1) ?? "no output"}`);
  }

  o.onProgress?.(1);
}

export interface ThumbnailOptions {
  durationMs: number;
  thumb: string;
  previews: string[];
  signal?: AbortSignal;
  onProgress?: (pct: number) => void;
}

/**
 * One cover frame (10% in — frame 0 is often black) plus `previews`, evenly spaced across the
 * rest of the recording. Idempotent per file, like extractPcm. A single frame that fails to
 * grab (e.g. a seek past a variable-framerate gap) is logged and skipped rather than failing
 * the whole job — thumbnails are not on the critical path the way transcription is.
 */
export async function extractThumbnails(
  cli: FfmpegCli,
  path: string,
  o: ThumbnailOptions,
): Promise<void> {
  throwIfAborted(o.signal);
  await mkdir(dirname(o.thumb), { recursive: true });

  const seconds = (f: number): number => (o.durationMs / 1000) * f;
  const targets: { out: string; at: number; width: number }[] = [
    { out: o.thumb, at: seconds(0.1), width: 320 },
    ...o.previews.map((out, i) => ({
      out,
      at: seconds((i + 1) / (o.previews.length + 1)),
      width: 160,
    })),
  ];

  let completed = 0;
  o.onProgress?.(0);
  for (const t of targets) {
    if (existsSync(t.out)) {
      completed += 1;
      o.onProgress?.(completed / targets.length);
      continue;
    }
    throwIfAborted(o.signal);
    try {
      await grabFrame(cli, path, t.at, t.width, t.out, o.signal);
    } catch (e) {
      if (isCancelled(e)) throw e;
      console.warn(`thumbnail frame failed for ${path} at ${t.at}s: ${String(e)}`);
    }
    completed += 1;
    o.onProgress?.(completed / targets.length);
  }
}

export async function grabFrame(
  cli: FfmpegCli,
  path: string,
  seconds: number,
  width: number,
  out: string,
  signal?: AbortSignal,
): Promise<void> {
  const tmp = `${out}.tmp`;
  const args = [
    "-nostdin", "-v", "error", "-y",
    "-ss", seconds.toFixed(3), "-i", path,
    "-frames:v", "1", "-vf", `scale=${width}:-2`, "-q:v", "4",
    // tmp name has no usable extension -> ffmpeg can't infer the muxer
    "-f", "image2", tmp,
  ];

  const result = await run(cli.ffmpeg, args, { signal });
  if (result.aborted) {
    discard(tmp);
    throw err.cancelled();
  }
  if (result.code !== 0) {
    discard(tmp);
    const lines = result.stderr.trim().split("\n").filter((l) => l !== "");
    throw err.ffmpeg(`${path}: ${lines.at(-1) ?? "no output"}`);
  }
  renameSync(tmp, out);
}

/**
 * ffmpeg's `-progress` stream is `key=value` lines; `out_time_us` is authoritative,
 * `out_time_ms` is a misnomer that also holds microseconds on older builds.
 */
export function progressMs(line: string): number | null {
  const eq = line.indexOf("=");
  if (eq < 0) return null;
  const key = line.slice(0, eq);
  if (key !== "out_time_us" && key !== "out_time_ms") return null;
  const value = line.slice(eq + 1).trim();
  // matches rust's u64 parse: "N/A" and ffmpeg's negative i64 sentinel are both rejected
  return /^\d+$/.test(value) ? Math.floor(Number(value) / 1000) : null;
}

/** A partial pcm is worse than none: the pipeline treats an existing file as a finished one. */
function discard(out: string): void {
  try {
    unlinkSync(out);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code !== "ENOENT") {
      console.warn(`could not remove partial pcm ${out}: ${String(e)}`);
    }
  }
}

interface RunResult {
  code: number;
  stdout: string;
  stderr: string;
  aborted: boolean;
}

/**
 * Both pipes are consumed as they arrive — a full stderr pipe blocks ffmpeg forever, which
 * is the deadlock the rust drains on a second thread to avoid.
 */
function run(
  bin: string,
  args: string[],
  o: { signal?: AbortSignal; onStdoutLine?: (line: string) => void } = {},
): Promise<RunResult> {
  return new Promise((resolve, reject) => {
    const child = spawn(bin, args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";
    let pending = "";
    let aborted = false;

    const onAbort = (): void => {
      aborted = true;
      child.kill();
    };
    if (o.signal) {
      if (o.signal.aborted) onAbort();
      else o.signal.addEventListener("abort", onAbort, { once: true });
    }

    child.stdout.setEncoding("utf8");
    child.stdout.on("data", (d: string) => {
      stdout += d;
      if (!o.onStdoutLine) return;
      pending += d;
      const lines = pending.split("\n");
      pending = lines.pop() ?? "";
      for (const line of lines) o.onStdoutLine(line);
    });
    child.stderr.setEncoding("utf8");
    child.stderr.on("data", (d: string) => {
      stderr += d;
    });

    child.on("error", (e: NodeJS.ErrnoException) => {
      o.signal?.removeEventListener("abort", onAbort);
      reject(e.code === "ENOENT" ? err.ffmpegMissing() : err.io(bin, e));
    });
    child.on("close", (code) => {
      o.signal?.removeEventListener("abort", onAbort);
      resolve({ code: code ?? -1, stdout, stderr, aborted });
    });
  });
}

function throwIfAborted(signal?: AbortSignal): void {
  if (signal?.aborted) throw err.cancelled();
}

const int = (v: string | undefined): number | null => {
  if (v === undefined) return null;
  const n = Number(v);
  return Number.isFinite(n) ? Math.trunc(n) : null;
};

interface RawProbe {
  streams?: RawStream[];
  format?: { format_name?: string; duration?: string };
}

interface RawStream {
  index: number;
  codec_type: string;
  codec_name?: string;
  channels?: number;
  sample_rate?: string;
  tags?: Record<string, string>;
  disposition?: { attached_pic?: number };
}
